### ISSUE TITLE
LLVM 8 segfault/Invalid PPC CTR loop! on ppc64le (again)

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#4026]
## Reporting a bug

Having updated LLVM support to LLVM8 https://github.com/numba/llvmlite/pull/478 and https://github.com/numba/numba/pull/4022 , it has become apparent that `ppc64le` builds based on this LLVM version are failing. Observations:

 * https://reviews.llvm.org/D53383 was added, which switched the relocation model for `ppc64le` from `PIC` to `Static`. Changing nothing in the Numba source leads to the default relocation model of `Static` being used and huge numbers of SIGSEGV occur when running the test suite (even for simple tests, and especially if `gufunc`s are used).
 * Changing the Numba source in this area:
  https://github.com/numba/numba/blob/e8ac4951affacd25c63ba2c18d62a3f12ed7e0ba/numba/targets/codegen.py#L793-L798
  so as to make it that `ppc64le` uses `PIC` relocation prevents the segfaults, but ends up signalling with `SIGABRT` from `UNREACHABLE executed at lib/Target/PowerPC/PPCCTRLoops.cpp:793:!` due to `Invalid PPC CTR loop!`. Which is the same thing that happened in LLVM 6.

**Comment by obilaniu** [comment]
## RuntimeDyldELF/TOC problem

Unfortunately, this seems to once again be a `RuntimeDyldELF`/TOC problem, but of a slightly different variant.

I selected an arbitrary segfaulting test, `numba.tests.test_alignment.TestAlignment.test_record_alignment` and ran it under

    gdb --args ~/.anaconda3/envs/numbadev/bin/python ./runtests.py numba.tests.test_alignment.TestAlignment.test_record_alignment

The segfault occurred within the compiled function, at a chunk that resembles the following:

       0x00003fffa2d202f4 <+444>:   bl      0x3fffa2d20434
       0x00003fffa2d202f8 <+448>:   ld      r2,0(r1)
    => 0x00003fffa2d202fc <+452>:   ld      r30,-32752(r3)
       0x00003fffa2d20300 <+456>:   mr      r3,r30

The return value of the function call at `bl` was `0x1` in `r3`, and it came from `PyArg_ParseTuple()`, indicating success. So I found it somewhat startling that the code would base a pointer off it, as the segfaulting `ld` seems to have done. Besides, the segfaulting `ld` has the "form" of a global access, which normally uses `r2`. And indeed, the TOC restoration clearly restores into `r2`, but the pointer access is made off `r3`. It's as if something is missing.

And indeed, something was missing. I immediately suspected the `ld r2,0(r1)` had trampled something very important underneath, and this hunch turned out to be correct. Compare the following versions, where the TOC restoration code is enabled and disabled respectively, both under the `static` relocation model:


### TOC restoration:

```
(gdb) disas
Dump of assembler code for function _ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE:
   0x00003fffa2d20138 <+0>:     ld      r2,-8(r12)
   0x00003fffa2d2013c <+4>:     add     r2,r2,r12
   0x00003fffa2d20140 <+8>:     mflr    r0
   0x00003fffa2d20144 <+12>:    std     r0,16(r1)
   0x00003fffa2d20148 <+16>:    stdu    r1,-176(r1)
   0x00003fffa2d2014c <+20>:    mr      r3,r4
   0x00003fffa2d20150 <+24>:    addis   r4,r2,0
   0x00003fffa2d20154 <+28>:    addi    r7,r1,152
   0x00003fffa2d20158 <+32>:    li      r5,1
   0x00003fffa2d2015c <+36>:    li      r6,1
   0x00003fffa2d20160 <+40>:    std     r30,160(r1)
   0x00003fffa2d20164 <+44>:    ld      r4,-32768(r4)
   0x00003fffa2d20168 <+48>:    bl      0x3fffa2d203dc
   0x00003fffa2d2016c <+52>:    ld      r2,24(r1)
   0x00003fffa2d20170 <+56>:    xxlxor  vs34,vs34,vs34
   0x00003fffa2d20174 <+60>:    cmplwi  r3,0
   0x00003fffa2d20178 <+64>:    li      r3,32
   0x00003fffa2d2017c <+68>:    addi    r4,r1,96
   0x00003fffa2d20180 <+72>:    li      r5,16
   0x00003fffa2d20184 <+76>:    stvx    v2,r4,r3
   0x00003fffa2d20188 <+80>:    li      r3,0
   0x00003fffa2d2018c <+84>:    stvx    v2,r4,r5
   0x00003fffa2d20190 <+88>:    stvx    v2,0,r4
   0x00003fffa2d20194 <+92>:    std     r3,144(r1)
   0x00003fffa2d20198 <+96>:    beq     0x3fffa2d20350 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+536>
   0x00003fffa2d2019c <+100>:   addis   r3,r2,0
   0x00003fffa2d201a0 <+104>:   ld      r3,-32760(r3)
   0x00003fffa2d201a4 <+108>:   ld      r3,0(r3)
   0x00003fffa2d201a8 <+112>:   cmpldi  r3,0
   0x00003fffa2d201ac <+116>:   beq     0x3fffa2d20324 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+492>
   0x00003fffa2d201b0 <+120>:   li      r3,32
   0x00003fffa2d201b4 <+124>:   li      r5,16
   0x00003fffa2d201b8 <+128>:   stvx    v2,0,r4
   0x00003fffa2d201bc <+132>:   stvx    v2,r4,r3
   0x00003fffa2d201c0 <+136>:   ld      r3,152(r1)
   0x00003fffa2d201c4 <+140>:   stvx    v2,r4,r5
   0x00003fffa2d201c8 <+144>:   li      r5,0
   0x00003fffa2d201cc <+148>:   std     r5,144(r1)
   0x00003fffa2d201d0 <+152>:   bl      0x3fffa2d20408
   0x00003fffa2d201d4 <+156>:   ld      r2,24(r1)
   0x00003fffa2d201d8 <+160>:   ld      r4,120(r1)
   0x00003fffa2d201dc <+164>:   cmpwi   r3,0
   0x00003fffa2d201e0 <+168>:   cmpdi   cr1,r4,16
   0x00003fffa2d201e4 <+172>:   crand   4*cr5+lt,eq,4*cr1+eq
   0x00003fffa2d201e8 <+176>:   bge     cr5,0x3fffa2d20338 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+512>
   0x00003fffa2d201ec <+180>:   ld      r9,112(r1)
   0x00003fffa2d201f0 <+184>:   ld      r3,96(r1)
   0x00003fffa2d201f4 <+188>:   cmpdi   r9,1
   0x00003fffa2d201f8 <+192>:   blt     0x3fffa2d202f4 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+444>
   0x00003fffa2d201fc <+196>:   addi    r4,r9,-1
   0x00003fffa2d20200 <+200>:   ld      r6,128(r1)
   0x00003fffa2d20204 <+204>:   clrldi  r5,r9,61
   0x00003fffa2d20208 <+208>:   li      r7,0
   0x00003fffa2d2020c <+212>:   cmpldi  r4,7
   0x00003fffa2d20210 <+216>:   ld      r4,144(r1)
   0x00003fffa2d20214 <+220>:   blt     0x3fffa2d202bc <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+388>
   0x00003fffa2d20218 <+224>:   rldicr  r8,r4,3,60
   0x00003fffa2d2021c <+228>:   subf    r9,r5,r9
   0x00003fffa2d20220 <+232>:   li      r7,0
   0x00003fffa2d20224 <+236>:   li      r10,8
   0x00003fffa2d20228 <+240>:   mr      r11,r6
   0x00003fffa2d2022c <+244>:   nop
   0x00003fffa2d20230 <+248>:   lfdx    f0,r11,r10
   0x00003fffa2d20234 <+252>:   add     r12,r4,r11
   0x00003fffa2d20238 <+256>:   addi    r7,r7,8
   0x00003fffa2d2023c <+260>:   add     r30,r4,r12
   0x00003fffa2d20240 <+264>:   cmpld   r9,r7
   0x00003fffa2d20244 <+268>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20248 <+272>:   stfiwx  f0,0,r11
   0x00003fffa2d2024c <+276>:   lfdx    f0,r12,r10
   0x00003fffa2d20250 <+280>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20254 <+284>:   stfiwx  f0,r4,r11
   0x00003fffa2d20258 <+288>:   add     r11,r11,r8
   0x00003fffa2d2025c <+292>:   lfdx    f0,r30,r10
   0x00003fffa2d20260 <+296>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20264 <+300>:   stfiwx  f0,r4,r12
   0x00003fffa2d20268 <+304>:   add     r12,r4,r30
   0x00003fffa2d2026c <+308>:   lfdx    f0,r12,r10
   0x00003fffa2d20270 <+312>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20274 <+316>:   stfiwx  f0,r4,r30
   0x00003fffa2d20278 <+320>:   add     r30,r4,r12
   0x00003fffa2d2027c <+324>:   lfdx    f0,r30,r10
   0x00003fffa2d20280 <+328>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20284 <+332>:   stfiwx  f0,r4,r12
   0x00003fffa2d20288 <+336>:   add     r12,r4,r30
   0x00003fffa2d2028c <+340>:   lfdx    f0,r12,r10
   0x00003fffa2d20290 <+344>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20294 <+348>:   stfiwx  f0,r4,r30
   0x00003fffa2d20298 <+352>:   add     r30,r4,r12
   0x00003fffa2d2029c <+356>:   lfdx    f0,r30,r10
   0x00003fffa2d202a0 <+360>:   xscvdpsxws vs0,vs0
   0x00003fffa2d202a4 <+364>:   stfiwx  f0,r4,r12
   0x00003fffa2d202a8 <+368>:   add     r12,r4,r30
   0x00003fffa2d202ac <+372>:   lfdx    f0,r12,r10
   0x00003fffa2d202b0 <+376>:   xscvdpsxws vs0,vs0
   0x00003fffa2d202b4 <+380>:   stfiwx  f0,r4,r30
   0x00003fffa2d202b8 <+384>:   bne     0x3fffa2d20230 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+248>
   0x00003fffa2d202bc <+388>:   cmpldi  r5,0
   0x00003fffa2d202c0 <+392>:   beq     0x3fffa2d202f4 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+444>
   0x00003fffa2d202c4 <+396>:   mulld   r7,r7,r4
   0x00003fffa2d202c8 <+400>:   mtctr   r5
   0x00003fffa2d202cc <+404>:   li      r5,8
   0x00003fffa2d202d0 <+408>:   add     r6,r6,r7
   0x00003fffa2d202d4 <+412>:   nop
   0x00003fffa2d202d8 <+416>:   nop
   0x00003fffa2d202dc <+420>:   nop
   0x00003fffa2d202e0 <+424>:   lfdx    f0,r6,r5
   0x00003fffa2d202e4 <+428>:   xscvdpsxws vs0,vs0
   0x00003fffa2d202e8 <+432>:   stfiwx  f0,0,r6
   0x00003fffa2d202ec <+436>:   add     r6,r6,r4
   0x00003fffa2d202f0 <+440>:   bdnz    0x3fffa2d202e0 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+424>
   0x00003fffa2d202f4 <+444>:   bl      0x3fffa2d20434
   0x00003fffa2d202f8 <+448>:   ld      r2,0(r1)
=> 0x00003fffa2d202fc <+452>:   ld      r30,-32752(r3)
   0x00003fffa2d20300 <+456>:   mr      r3,r30
   0x00003fffa2d20304 <+460>:   bl      0x3fffa2d20460
   0x00003fffa2d20308 <+464>:   ld      r2,24(r1)
   0x00003fffa2d2030c <+468>:   mr      r3,r30
   0x00003fffa2d20310 <+472>:   ld      r30,160(r1)
   0x00003fffa2d20314 <+476>:   addi    r1,r1,176
   0x00003fffa2d20318 <+480>:   ld      r0,16(r1)
   0x00003fffa2d2031c <+484>:   mtlr    r0
   0x00003fffa2d20320 <+488>:   blr
   0x00003fffa2d20324 <+492>:   addis   r3,r2,0
   0x00003fffa2d20328 <+496>:   addis   r4,r2,0
   0x00003fffa2d2032c <+500>:   ld      r3,-32744(r3)
   0x00003fffa2d20330 <+504>:   ld      r4,-32736(r4)
   0x00003fffa2d20334 <+508>:   b       0x3fffa2d20348 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+528>
   0x00003fffa2d20338 <+512>:   addis   r3,r2,0
   0x00003fffa2d2033c <+516>:   addis   r4,r2,0
   0x00003fffa2d20340 <+520>:   ld      r3,-32728(r3)
   0x00003fffa2d20344 <+524>:   ld      r4,-32720(r4)
   0x00003fffa2d20348 <+528>:   bl      0x3fffa2d2048c
   0x00003fffa2d2034c <+532>:   ld      r2,24(r1)
   0x00003fffa2d20350 <+536>:   li      r3,0
   0x00003fffa2d20354 <+540>:   b       0x3fffa2d20310 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+472>
   0x00003fffa2d20358 <+544>:   .long 0x0
   0x00003fffa2d2035c <+548>:   .long 0x0
   0x00003fffa2d20360 <+552>:   .long 0x0
End of assembler dump.
```


### No TOC restoration:

```
(gdb) disas
Dump of assembler code for function _ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE:
   0x00003fffa2d20138 <+0>:     ld      r2,-8(r12)
   0x00003fffa2d2013c <+4>:     add     r2,r2,r12
   0x00003fffa2d20140 <+8>:     mflr    r0
   0x00003fffa2d20144 <+12>:    std     r0,16(r1)
   0x00003fffa2d20148 <+16>:    stdu    r1,-176(r1)
   0x00003fffa2d2014c <+20>:    mr      r3,r4
   0x00003fffa2d20150 <+24>:    addis   r4,r2,0
   0x00003fffa2d20154 <+28>:    addi    r7,r1,152
   0x00003fffa2d20158 <+32>:    li      r5,1
   0x00003fffa2d2015c <+36>:    li      r6,1
   0x00003fffa2d20160 <+40>:    std     r30,160(r1)
   0x00003fffa2d20164 <+44>:    ld      r4,-32768(r4)
   0x00003fffa2d20168 <+48>:    bl      0x3fffa2d203dc
   0x00003fffa2d2016c <+52>:    nop
   0x00003fffa2d20170 <+56>:    xxlxor  vs34,vs34,vs34
   0x00003fffa2d20174 <+60>:    cmplwi  r3,0
   0x00003fffa2d20178 <+64>:    li      r3,32
   0x00003fffa2d2017c <+68>:    addi    r4,r1,96
   0x00003fffa2d20180 <+72>:    li      r5,16
   0x00003fffa2d20184 <+76>:    stvx    v2,r4,r3
   0x00003fffa2d20188 <+80>:    li      r3,0
   0x00003fffa2d2018c <+84>:    stvx    v2,r4,r5
   0x00003fffa2d20190 <+88>:    stvx    v2,0,r4
   0x00003fffa2d20194 <+92>:    std     r3,144(r1)
   0x00003fffa2d20198 <+96>:    beq     0x3fffa2d20350 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+536>
   0x00003fffa2d2019c <+100>:   addis   r3,r2,0
   0x00003fffa2d201a0 <+104>:   ld      r3,-32760(r3)
=> 0x00003fffa2d201a4 <+108>:   ld      r3,0(r3)
   0x00003fffa2d201a8 <+112>:   cmpldi  r3,0
   0x00003fffa2d201ac <+116>:   beq     0x3fffa2d20324 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+492>
   0x00003fffa2d201b0 <+120>:   li      r3,32
   0x00003fffa2d201b4 <+124>:   li      r5,16
   0x00003fffa2d201b8 <+128>:   stvx    v2,0,r4
   0x00003fffa2d201bc <+132>:   stvx    v2,r4,r3
   0x00003fffa2d201c0 <+136>:   ld      r3,152(r1)
   0x00003fffa2d201c4 <+140>:   stvx    v2,r4,r5
   0x00003fffa2d201c8 <+144>:   li      r5,0
   0x00003fffa2d201cc <+148>:   std     r5,144(r1)
   0x00003fffa2d201d0 <+152>:   bl      0x3fffa2d20408
   0x00003fffa2d201d4 <+156>:   nop
   0x00003fffa2d201d8 <+160>:   ld      r4,120(r1)
   0x00003fffa2d201dc <+164>:   cmpwi   r3,0
   0x00003fffa2d201e0 <+168>:   cmpdi   cr1,r4,16
   0x00003fffa2d201e4 <+172>:   crand   4*cr5+lt,eq,4*cr1+eq
   0x00003fffa2d201e8 <+176>:   bge     cr5,0x3fffa2d20338 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+512>
   0x00003fffa2d201ec <+180>:   ld      r9,112(r1)
   0x00003fffa2d201f0 <+184>:   ld      r3,96(r1)
   0x00003fffa2d201f4 <+188>:   cmpdi   r9,1
   0x00003fffa2d201f8 <+192>:   blt     0x3fffa2d202f4 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+444>
   0x00003fffa2d201fc <+196>:   addi    r4,r9,-1
   0x00003fffa2d20200 <+200>:   ld      r6,128(r1)
   0x00003fffa2d20204 <+204>:   clrldi  r5,r9,61
   0x00003fffa2d20208 <+208>:   li      r7,0
   0x00003fffa2d2020c <+212>:   cmpldi  r4,7
   0x00003fffa2d20210 <+216>:   ld      r4,144(r1)
   0x00003fffa2d20214 <+220>:   blt     0x3fffa2d202bc <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+388>
   0x00003fffa2d20218 <+224>:   rldicr  r8,r4,3,60
   0x00003fffa2d2021c <+228>:   subf    r9,r5,r9
   0x00003fffa2d20220 <+232>:   li      r7,0
   0x00003fffa2d20224 <+236>:   li      r10,8
   0x00003fffa2d20228 <+240>:   mr      r11,r6
   0x00003fffa2d2022c <+244>:   nop
   0x00003fffa2d20230 <+248>:   lfdx    f0,r11,r10
   0x00003fffa2d20234 <+252>:   add     r12,r4,r11
   0x00003fffa2d20238 <+256>:   addi    r7,r7,8
   0x00003fffa2d2023c <+260>:   add     r30,r4,r12
   0x00003fffa2d20240 <+264>:   cmpld   r9,r7
   0x00003fffa2d20244 <+268>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20248 <+272>:   stfiwx  f0,0,r11
   0x00003fffa2d2024c <+276>:   lfdx    f0,r12,r10
   0x00003fffa2d20250 <+280>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20254 <+284>:   stfiwx  f0,r4,r11
   0x00003fffa2d20258 <+288>:   add     r11,r11,r8
   0x00003fffa2d2025c <+292>:   lfdx    f0,r30,r10
   0x00003fffa2d20260 <+296>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20264 <+300>:   stfiwx  f0,r4,r12
   0x00003fffa2d20268 <+304>:   add     r12,r4,r30
   0x00003fffa2d2026c <+308>:   lfdx    f0,r12,r10
   0x00003fffa2d20270 <+312>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20274 <+316>:   stfiwx  f0,r4,r30
   0x00003fffa2d20278 <+320>:   add     r30,r4,r12
   0x00003fffa2d2027c <+324>:   lfdx    f0,r30,r10
   0x00003fffa2d20280 <+328>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20284 <+332>:   stfiwx  f0,r4,r12
   0x00003fffa2d20288 <+336>:   add     r12,r4,r30
   0x00003fffa2d2028c <+340>:   lfdx    f0,r12,r10
   0x00003fffa2d20290 <+344>:   xscvdpsxws vs0,vs0
   0x00003fffa2d20294 <+348>:   stfiwx  f0,r4,r30
   0x00003fffa2d20298 <+352>:   add     r30,r4,r12
   0x00003fffa2d2029c <+356>:   lfdx    f0,r30,r10
   0x00003fffa2d202a0 <+360>:   xscvdpsxws vs0,vs0
   0x00003fffa2d202a4 <+364>:   stfiwx  f0,r4,r12
   0x00003fffa2d202a8 <+368>:   add     r12,r4,r30
   0x00003fffa2d202ac <+372>:   lfdx    f0,r12,r10
   0x00003fffa2d202b0 <+376>:   xscvdpsxws vs0,vs0
   0x00003fffa2d202b4 <+380>:   stfiwx  f0,r4,r30
   0x00003fffa2d202b8 <+384>:   bne     0x3fffa2d20230 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+248>
   0x00003fffa2d202bc <+388>:   cmpldi  r5,0
   0x00003fffa2d202c0 <+392>:   beq     0x3fffa2d202f4 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+444>
   0x00003fffa2d202c4 <+396>:   mulld   r7,r7,r4
   0x00003fffa2d202c8 <+400>:   mtctr   r5
   0x00003fffa2d202cc <+404>:   li      r5,8
   0x00003fffa2d202d0 <+408>:   add     r6,r6,r7
   0x00003fffa2d202d4 <+412>:   nop
   0x00003fffa2d202d8 <+416>:   nop
   0x00003fffa2d202dc <+420>:   nop
   0x00003fffa2d202e0 <+424>:   lfdx    f0,r6,r5
   0x00003fffa2d202e4 <+428>:   xscvdpsxws vs0,vs0
   0x00003fffa2d202e8 <+432>:   stfiwx  f0,0,r6
   0x00003fffa2d202ec <+436>:   add     r6,r6,r4
   0x00003fffa2d202f0 <+440>:   bdnz    0x3fffa2d202e0 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+424>
   0x00003fffa2d202f4 <+444>:   bl      0x3fffa2d20434
   0x00003fffa2d202f8 <+448>:   addis   r3,r2,0
   0x00003fffa2d202fc <+452>:   ld      r30,-32752(r3)
   0x00003fffa2d20300 <+456>:   mr      r3,r30
   0x00003fffa2d20304 <+460>:   bl      0x3fffa2d20460
   0x00003fffa2d20308 <+464>:   nop
   0x00003fffa2d2030c <+468>:   mr      r3,r30
   0x00003fffa2d20310 <+472>:   ld      r30,160(r1)
   0x00003fffa2d20314 <+476>:   addi    r1,r1,176
   0x00003fffa2d20318 <+480>:   ld      r0,16(r1)
   0x00003fffa2d2031c <+484>:   mtlr    r0
   0x00003fffa2d20320 <+488>:   blr
   0x00003fffa2d20324 <+492>:   addis   r3,r2,0
   0x00003fffa2d20328 <+496>:   addis   r4,r2,0
   0x00003fffa2d2032c <+500>:   ld      r3,-32744(r3)
   0x00003fffa2d20330 <+504>:   ld      r4,-32736(r4)
   0x00003fffa2d20334 <+508>:   b       0x3fffa2d20348 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+528>
   0x00003fffa2d20338 <+512>:   addis   r3,r2,0
   0x00003fffa2d2033c <+516>:   addis   r4,r2,0
   0x00003fffa2d20340 <+520>:   ld      r3,-32728(r3)
   0x00003fffa2d20344 <+524>:   ld      r4,-32720(r4)
   0x00003fffa2d20348 <+528>:   bl      0x3fffa2d2048c
   0x00003fffa2d2034c <+532>:   nop
   0x00003fffa2d20350 <+536>:   li      r3,0
   0x00003fffa2d20354 <+540>:   b       0x3fffa2d20310 <_ZN7cpython5numba5tests14test_alignment13TestAlignment21test_record_alignment12$3clocals$3e7foo$241E5ArrayI6RecordILi321EELi1E1A7mutable7alignedE+472>
   0x00003fffa2d20358 <+544>:   .long 0x0
   0x00003fffa2d2035c <+548>:   .long 0x0
   0x00003fffa2d20360 <+552>:   .long 0x0
End of assembler dump.
```


## Significant Difference:

There are more instances of this, but the location of the crash changed very meaningfully:

### TOC restoration:

       0x00003fffa2d202f4 <+444>:   bl      0x3fffa2d20434
       0x00003fffa2d202f8 <+448>:   ld      r2,0(r1)
    => 0x00003fffa2d202fc <+452>:   ld      r30,-32752(r3)
       0x00003fffa2d20300 <+456>:   mr      r3,r30

### No TOC restoration:

       0x00003fffa2d202f4 <+444>:   bl      0x3fffa2d20434
       0x00003fffa2d202f8 <+448>:   addis   r3,r2,0
       0x00003fffa2d202fc <+452>:   ld      r30,-32752(r3)
       0x00003fffa2d20300 <+456>:   mr      r3,r30


So the compiler did not generate a `nop` after the `bl` that could serve as a TOC restoration slot. Rather, because the relocation model is `static` a and the call is intra-object, it appears that the compiler allowed itself to do "something useful" in that slot instead. But RuntimeDyldELF, not knowing that the compiler left it no nop slot and expecting one there, then unconditionally overwrote the important instruction `addis` immediately following the call.


## To Do

I suspect that my patch in #2451 may not have been sufficiently accurate. At a bare minimum, it appears that under the `static` relocation model, intra-object calls do not leave a ROC restoration slot afterwards. Therefore, rather than having the option of calling into the Global Entry Point (GEP) or not, these calls *must* use the Local Entry Point (LEP) and *must not* restore the TOC `r2` afterwards.

I had believed that it was always safe to conservatively call into GEP, but it turns out that there are times when it is strictly forbidden to call into one or the other.

My preference is for a greater expert than myself to mirror into `RuntimeDyldELF` the *exact* logic used by codegen to decide whether or not to 1) Call into the GEP, 2) Relax to LEP and 3) Whether a TOC restoration slot should be present or not. The possibility that there might not be enough information left at the RuntimeDyldELF stage slightly unnerves me.

Pinging @lhames since he is one of the knowledgeable PPC64 experts who contributed to the fix for #2451.

**Comment by lhames** [comment]
Hi Olexa,

You've oversold my PPC64 credentials, but for what it's worth your
reasoning sounds right to me.

Looking at your fix from #2451
<https://github.com/numba/numba/issues/2451> (committed
as llvm svn r329335), it looks like we compared SectionIDs to determine
whether to to restore TOCs:

if (IsExtern || (AbiVariant == 2 && Value.SectionID != SectionID))

But that would mean that calls between sections within the same object
would (incorrectly from the sounds of things) restore the TOC.

Does that match what you're seeing? Are these failures calls between
sections within a single object?

If that's the case, what happens if you replace the test (and any related
ones) with:
if (IsExtern ||
   (AbiVariant == 2 && Sections[Value.SectionID].ObjAddress !=
Sections[SectionID].ObjAddress))

That should expand the test from "is the callee in the same section as the
caller", to "is the callee in the same object as the caller".

Cheers,
Lang.

**Comment by stuartarchibald** [comment]
@obilaniu Many thanks for once again looking at the issues hitting PPC64LE support. I guess there are two items here.

1. The noted problem above with TOC etc for static relocation and RuntimeDyldELF.
2. Something unknown that has broken PIC relocation (Invalid CTR loop issue).

@lhames thanks also for contributing suggestions etc, much appreciated.

The Numba team has access to relevant hardware so we can help with testing ideas if that's useful.

Thanks again to you both!

**Comment by obilaniu** [comment]
@lhames Yes, the errors relate to intra-object calls, but unfortunately that proposed fix didn't work...

I really think we must find and either reuse or (less ideally) copy the exact logic that codegen used in the first place to decide or not to restore TOC.

**Comment by nemanjai** [comment]
The function in the PPC back end that decides which functions share a TOC (and will therefore be called through the Local Entry Point) is `callsShareTOCBase()`. There are a few possible solutions here:
- Perhaps we can expose this function through TargetLoweringInfo and use it in the dynamic linker
- We can replicate the logic in the dynamic linker
- We can provide a mode where every non-leaf function will save the TOC and every call will be followed by a `nop` so the TOC restore can be added (i.e. completely pessimistic in terms of TOC)

**Comment by stuartarchibald** [comment]
Thanks for looking at this @nemanjai. IIRC the issue still exists in LLVM 9, we've just patched the Numba test suite to not run tests that trigger it, decorator is `@skip_ppc64le_issue4026`. If it'd help get this fixed upstream we can try and pull some reproducing LLVM IR out of the tests that are triggering the problem. Thanks again!

**Comment by jayfurmanek** [comment]
If we can get reproducing LLVM IR, I think that would certainly help.

**Comment by edelsohn** [comment]
@obilaniu What level of bounty would entice you to fix this issue?

**Comment by n5ya7e** [comment]
@sklam I believe that this issue was closed prematurely because it actually describes two issues.  One is the Invalid CTR issue and the other is the RuntimeDyldELF issue.  The Invalid CTR issue was fixed in LLVM, but the RuntimeDyldELF issue remains.

**Comment by edelsohn** [comment]
@sklam I believe that this issue was closed prematurely.  The Invalid CTR bug was fixed in LLVM, but not the RuntimeDyldELF/TOC bug reported in the comment.

**Comment by stuartarchibald** [comment]
Reopening issue for TOC bug.

**Comment by resistor** [comment]
@edelsohn I don't think it makes sense to keep this bug open, as it does not reproduce with current Numba or LLVM.  Without a reproducing case, there's nothing actionable.

**Comment by edelsohn** [comment]
@resistor The problem depends on whether Numba is built with or without PIC.  If Numba has reverted back to PIC, it will not show the bug, but it remains latent.

If Numba wants to close this issue, okay, but the focus was the RuntimeDyldELF bug that remains unresolved.  The bounty is associated with the RuntimeDyldELF facet of the issue.

**Comment by resistor** [comment]
@edelsohn I tested with and without PIC and did not see the issue reproduce.

**Comment by edelsohn** [comment]
If Numba functionality has been restored on PPC64LE, that's great.  I assume, based on the comments in the pull request, that Numba now using LLVM 10.

We at IBM are concerned that there is a latent bug in RuntimeDyldELF for PPC64LE, but apparently there is no reproducer.

If you want to declare this issue fixed, that's fine.  But it's not clear that the TOC issue truly was fixed.  The PPC CTR loop issue was known to be fixed upstream in LLVM by the IBM team.

**Comment by resistor** [comment]
I have a candidate patch for the TOC issue, but I don't have a way to test it without a test that confirms the fix.  I too would like to land it in LLVM, but I can't with confirming that it actually resolves the issue.

**Comment by edelsohn** [comment]
Thanks for investigating the TOC issue.

Can you test the patch with Numba and LLVM 8, which seemed to reproduce the issue, according to Obilaniu?

**Comment by resistor** [comment]
Maybe.  Mainline Numba no longer supports LLVM 8, so it will be a pain to setup an environment to test it.

**Comment by edelsohn** [comment]
Can you test your LLVM change with the Numba release that was available in April 2019 and LLVM 8?  That seemed to be the combination that reproduced the problem.

**Comment by abebeos** [comment]
@resistor , may I ask what the status here is?

**Comment by resistor** [comment]
> @resistor , may I ask what the status here is?

I ran out of time to investigate it.

**Comment by abebeos** [comment]
I understand, bounties are sometimes "never ending".

**Comment by abebeos** [comment]
> If Numba wants to close this issue, okay, but the focus was the RuntimeDyldELF bug that remains unresolved. The bounty is associated with the RuntimeDyldELF facet of the issue.

> We at IBM are concerned that there is a latent bug in RuntimeDyldELF for PPC64LE, but apparently there is no reproducer.

@edelsohn, what about a new (related) bounty of the kind:

* Create test-environment for the issue #4026 
* Should allow ideally to work on the relevant sources (llvmlite, llvm, numba)
* Should work ideally on non-ppc hardware, too

My personal experience tells me that a "ready-to-go dev environment" can do wonders, simply because an interested developer can start right away without draining energy in setups.

Additionally, I noticed within several ppc64le-issues/bounties the problem "I don't have hardware", so possibly this should be solved, too (reusable for other ppc64le issues).

**Comment by lhames** [comment]
@edelsohn @obilaniu @resistor Unfortunately I don't have time to work on this myself, but if you find someone who does please tell them that I will be happy to answer general RuntimeDyld questions.

If Numba uses LLVM's OrcV2 APIs (either via LLJIT, or a custom class) then I would also recommend that they consider writing a JITLink implementation for PPC, rather than just patching RuntimeDyld. JITLink is RuntimeDyld's successor and provides richer data structures and a cleaner design than RuntimeDyld. Writing a JITLink implementation would be a little more work but is likely to be more maintainable in the long term, and may result in better performance (I don't recall the details for PPC, but RuntimeDyld generally takes shortcuts and relies on accesses being indirect via registers. JITLink uses the native code model with direct accesses where possible). This may make a good Google Summer Of Code project if you find a motivated student.

**Comment by abebeos** [comment]
I'm on this since a few days, staying on it during the holidays. Would have preferred to have the reproducer/dev-env as separate chunk-of-work, but ok, started anyways. Will update here if I have news or questions. @lhames , pinging you here would be ok for you?

**Comment by lhames** [comment]
@abebeos Sure. I'm traveling though, so replies may be delayed.

**Comment by abebeos** [comment]
From different comments:

>@edelsohn I don't think it makes sense to keep this bug open, as it does not reproduce with current Numba or LLVM. Without a reproducing case, there's nothing actionable.

> Can you test your LLVM change with the Numba release that was available in April 2019 and LLVM 8? That seemed to be the combination that reproduced the problem.

> Maybe. Mainline Numba no longer supports LLVM 8, so it will be a pain to setup an environment to test it.

Good News:

* I "went through the pain" (numba really needs a core-dev-docker), made several setups with different versions, and reproduced the bug in several of them. 
* Went finally with the latest versions (llvm10, llvmlite v0.35.0 numba 0.52.0).
* Attempt to fix on RuntimeDyld side failed, it felt anyways not right (sidenote: code would need a refactoring-for-clarity)
* Moved over to compiler/Target side, after some work-in and research, identified fix.
* => the few selected segfaulting tests work now (both, pic/static).
* => full test-suite currently running (both, pic/static)
* => even if this fails: very confident that it'll work in two or three days.

(Posting more details and patches / test-sequence after getting some sleep.)

**Comment by edelsohn** [comment]
Great!  I'm glad that you were able to identify a specific problem.  Thanks for the investigation!  I look forward to feedback from @nemanjai .

I'm not certain if refactoring RuntimeDyld would be useful or better to re-write it in the new JITLink infrastructure. I would be interested to receive price estimates for a JITLink implementation.

**Comment by abebeos** [comment]
> I'm not certain if refactoring RuntimeDyld would be useful or better to re-write it in the new JITLink infrastructure. 

I couldn't understand RtDyld, so I refactored the relevant sections a bit. After understanding, it became clear to move over to the compiler. The (partial) refactoring (strict, mechanical) would be not that much effort (possibly similar to this issue/bounty), and possibly anyways necessary, so there is a better foundation to do a JITLink implementation (assuming that one would always peek back to RtDyld).

> I would be interested to receive price estimates for a JITLink implementation.

Ok, just give me a few more days to ensure the fix works, and then I can get a picture of the overall JITLink problem-domain. My (strong) opinion is that a few organizational tasks should be done prior to going over to a greater code-tasks like JITLink.

I'll added a section to https://github.com/numba/numba/issues/6606#rca-further-tasks to keep track.

**Comment by abebeos** [comment]
@lhames , please take a look at #6606 and related patch #684 - patch send to https://reviews.llvm.org/D94183, it should be trivial for you to get this in (it's a small hell for me)...

**Comment by lhames** [comment]
Hi Abebeos,

That patch looks like it's for LLVM 10.x, but older LLVM versions are not
maintained.

This patch should be rebased on the main branch and a review submitted as
soon as possible so that it can be included in LLVM 12 which is branching
soon. You could also ask whether it can be included in the upcoming LLVM
11.0.1 release, but I'm not sure whether they're still accepting changes.
You will need a PPC maintainer to review this change as it's in the PPC
backend.

-- Lang.

**Comment by edelsohn** [comment]
I believe that the proposed patch is a backport of a patch already in current LLVM.  The development team wasn't aware that the change affected Numba.  Thanks for confirming that the patch fixes the problem and there is no latent bug.

If that is the correct assessment, IBM is willing to award the bounty for confirming that the change already upstream fixed the problem observed in Numba.

**Comment by lhames** [comment]
Hi David,

I can not confirm the fix as I'm not a PPC backend expert, and am currently
busy working on bringing up the new JIT linker, I was only speaking to the
review process for getting it in to LLVM.

I will be happy to help with RuntimeDyld reviews though, and would be happy
to help Numba port their code to LLVM's newer OrcV2 / JITLink APIs once we
have a PPC implementation of JITLink.

-- Lang.

**Comment by abebeos** [comment]
> Hi David, I can not confirm the fix as I'm not a PPC backend expert, and am currently busy working on bringing up the new JIT linker,
[...]

You mean for PowerPC? That JITLink that replaces RtDyldELF?

**Comment by abebeos** [comment]
> I believe that the proposed patch is a backport of a patch already in current LLVM. The development team wasn't aware that the change affected Numba. Thanks for confirming that the patch fixes the problem and there is no latent bug.

Yes, its already in current LLVM, see: https://reviews.llvm.org/rG2812c1515627904e31605bbd4f25a887a1f8eb12

I "just" identified it, backported and validated in older numba setup.
 
> If that is the correct assessment, IBM is willing to award the bounty for confirming that the change already upstream fixed the problem observed in Numba.

Very nice. I would suggest to close this issue here ,and keep #6606 open until the patch is merged into llvmlite 0.54s (patches llvm11, see https://github.com/numba/numba/issues/6606#issuecomment-755396144 )

**Comment by stuartarchibald** [comment]
>  I will be happy to help with RuntimeDyld reviews though, and would be happy to help Numba port their code to LLVM's newer OrcV2 / JITLink APIs once we have a PPC implementation of JITLink.

@lhames thanks for offering to help with this port, much appreciated. I'm not sure yet when this change will be scheduled but will make sure you are pinged on the associated issue/PR once work starts. Thanks again!

**Comment by edelsohn** [comment]
@abebeos seems to have confirmed the problem and the solution.  I consider this issue resolved.

@stuartarchibald Can you close this issue, if you agree?

Proposals to implement JITLink for PPC are welcome.

**Comment by abebeos** [comment]
> @abebeos seems to have confirmed the problem and the solution. I consider this issue resolved.
>
> @stuartarchibald Can you close this issue, if you agree?

@esc can you please try to reach @stuartarchibald - or close it by yourself?

The issue was anyways reopened for the TOC bug, see https://github.com/numba/numba/issues/4026#issuecomment-704478230, which is now tracked within #6066). 

(don't want to be "pushy" here, its just that Thursday is a special weekly-deadline-day at Bountysource, and there were already many delays due to the holidays)

> Proposals to implement JITLink for PPC are welcome.

@edelsohn, I have one here https://github.com/abebeos/PowerPC/issues/4  (hopefully I've a concrete estimate in two or three days).

**Comment by esc** [comment]
Dear @abebeos and @edelsohn -- thank you for asking about closing this issue. I conferred OOB with @stuartarchibald and @sklam on this and will report our stance here now. We are happy to close this issue for now, but we---the maintainers of the Numba project--do not consider this issue 'resolved' per se. The Numba (and hence also llvmlite) policy on considering something as 'resolved' is once a fix has been merged to `mainline` (aka `master`) and a development build becomes available.

In this case  however, as @abebeos mentions, this particular issue can be closed as it being tracked by #6606 (btw. not #6066 as mentioned above, simple typo) https://github.com/numba/llvmlite/pull/684. As communicated in https://github.com/numba/numba/issues/6606#issuecomment-755396144 and https://github.com/numba/numba/issues/6606#issuecomment-757822179 we will attempt to resolve issue #6606 as part of the upgrade to LLVM in due course. The current best guess for moving forward, will be to port the existing patch from https://github.com/numba/llvmlite/pull/684 to LLVM 11.* after we upgrade llvmlite to support LLVM 11.0 and the run it through our internal build farm with a version of Numba that has the correct reolc model configured to definitively confirm the fix from our end. At that point and everything goes well, we will close #6606 and merge https://github.com/numba/llvmlite/pull/684.

We hope that this course of action is deemed acceptable for all parties involved. If this is not the case, please do voice your concerns and suggest an alternative course of action which we will be happy to consider and discuss.

Thank you all!

**Comment by edelsohn** [comment]
@stuartarchibald @esc I'm confused.  I thought that this problem already was fixed, or at least latent, in Numba and the only issue was confirming the patch solved the latent issue in the earlier release of Numba.

I am not tracking the issue separately in #6606 and that seems to be a duplicate of the latent issue.  I don't consider #6606 an open issue either.

Please clarify why this isn't the case.

**Comment by abebeos** [comment]
## #4026 LLVM 8 segfault/Invalid PPC CTR loop! on ppc64le (again) 

* Solved by IBM Team (not sure, possibly by this: https://reviews.llvm.org/D81126?id=274770)
* Issue was closed by numba-team (merge of #6300, re-enabling tests)
* Issue was reopened for the (latent) TOC-bug by https://github.com/numba/numba/issues/4026#issuecomment-704478230 at request https://github.com/numba/numba/issues/4026#issuecomment-704429053

## #4026 TOC-bug in numba

* was fixed, by changing reloc_mode to pic (https://github.com/numba/numba/commit/31ddd81d559c4aa79fb25e4c7b45b44451df95d5)

Thus, #4026 is fixed/closed anyways (both issues, CTR loop & TOC are merged)

## #6606 TOC-bug, latent

* I followed standard-procedure, opened a separate issue
* Standard-Procedure (which numba-team follows) is to keep it open until merged

(sidenote: llvm-12 contains all fixes)

-

All is good, possibly just a misunderstanding.

**Comment by edelsohn** [comment]
@abebeos Please don't claim the bounty until we have consensus.  I am going to reject the bounty request.

**Comment by abebeos** [comment]
> @abebeos Please don't claim the bounty until we have consensus. I am going to reject the bounty request.

???

Claimed it already, as it seemed like a technicality after this:

> I believe that the proposed patch is a backport of a patch already in current LLVM. The development team wasn't aware that the change affected Numba. Thanks for confirming that the patch fixes the problem and there is no latent bug.
> 
> If that is the correct assessment, IBM is willing to award the bounty for confirming that the change already upstream fixed the problem observed in Numba.

> @abebeos seems to have confirmed the problem and the solution. I consider this issue resolved.

Just don't ACK it, no need to reject.

**Comment by edelsohn** [comment]
I would like to have consensus about the status.

The bounty was opened against this issue because I thought that this problem had resurfaced. And this issue had become a meta-issue for the various issues. I asked for this to be closed because I thought that the issue was completely resolved, not close this issue but move the dependency to #6606 because that causes more disconnect.

My understanding is:
1. The symptom is not present in current Numba.
2. The latent problem has been fixed in LLVM.
3. The patch proposed backporting the fix to LLVM 10.0 to solve the problem in the previous release of Numba, but that will not occur because of LLVM support policies. Backporting the patch cannot be a requirement to resolve the issue.
4. llvmlite should to be updated to the current version of LLVM that contains the fix, and Numba should utilize the updated llvmlite.  But because the bug currently is latent, this should have no effect on Numba and only should ensure that the bug does not reappear.

**Comment by abebeos** [comment]
[...]
> My understanding is:
> 
>     1. The symptom is not present in current Numba.

correct (as it uses reloc_mode=pic for ppc)

>     2. The latent problem has been fixed in LLVM.

correct
 
>     3. The patch proposed backporting the fix to LLVM 10.0 to solve the problem in the previous release of Numba, but that will not occur because of LLVM support policies. Backporting the patch cannot be a requirement to resolve the issue.

agree
 
>     4. llvmlite should to be updated to the current version of LLVM that contains the fix, and Numba should utilize the updated llvmlite.  But because the bug currently is latent, this should have no effect on Numba and only should ensure that the bug does not reappear.

correct.

**Comment by abebeos** [comment]
>  to port the existing patch from [numba/llvmlite#684](https://github.com/numba/llvmlite/pull/684) to LLVM 11.*

I'll provide the PR tomorrow, so you have it available when needed.

**Comment by esc** [comment]
Hi @edelsohn and @abebeos -- I will now try to summarize the current state of this as far as I understand it.

This issue originally referred to two  bugs: "The Invalid CTR bug" which  was fixed in LLVM and the "RuntimeDyldELF/TOC" for which the issue was re-opened as @abebeos mentioned above:

> Issue was reopened for the (latent) TOC-bug by #4026 (comment) at request #4026 (comment)

The issue is dormant or latent, because it will not manifest under the current relocation model (`pic` not `static` which is what Numba currently uses for PPC. This was also correctly identified by @abebeos 

> was fixed, by changing reloc_mode to pic (31ddd81)

Also note, that patch to fix this is already present in LLVM mainline but not 10.* and 11.*  and LLVM policies prevent inclusion of a backport (see also: https://reviews.llvm.org/D94183). This is also correct and has been explained by @abebeos above:

> Yes, its already in current LLVM, see: https://reviews.llvm.org/rG2812c1515627904e31605bbd4f25a887a1f8eb12

However, I should mention, that the term 'fixed' is somewhat ambiguous in this context. The bug would manifest again, if Numba would change the relocation model back to `static` right now. It is not yet clear to me however, if this would have any benefit in terms of usability or performance. Perhaps someone on this thread could share some insight on this? Also, from the Numba side this isn't a high-priority issue, since the conclusion is that using the `pic` reloc model on Power will be fine for the time being, i.e. there is nothing "immediately on fire".

Thus, in order for Numba to change the relocation model, the latent TOC issue should be addressed. Specifically it should be addressed in llvmlite. Now, there are multiple routes to addressing, which are coupled strongly to the LLVM version in use. Note that currently, for PPC, llvmlite supports only LLVM 10.*.

Option A: merge the existing patch by @abebeos for LLVM 10.* and rebuild `llvmlite`.

Option B: first port `llvmlite` to LLVM 11.* -- then port the patch (or use the ported patch if it becomes available before that) and then rebuild llvmlite to be based on LLVM 11.* incl. the patch. 

Option C: port `llvmlite` to LLVM 12.*  which includes the patch (specifically a LLVM 12.* version of it) already and then rebuild `llvmlite` to be based on LLVM 12.*. Since LLVM 12.*  has not been released yet and we would like to wait for LLVM 12.0.1 (first bug fix release) anyway before upgrading llvmlite, this is something we can aim to do during the second half of 2021 or maybe not even until 2022.

If any one of the three options above are in place, it would be feasible to switch the relocation model in Numba for PPC from `pic` to `static` again and reap any potential benefits that result from this change. 

During the developer meeting at the start of Jan (https://github.com/numba/numba/wiki/Minutes_2021_01_05)  we concluded to attempt Option B for Numba 0.54. We are currently in the midst of wrapping things up for Numba 0.53 release and hope to get a release candidate (RC) out by end of Jan / beginning of Feb. Once the RC is out, there will be capacity (hardware, software, human) to embark on Option B as I have communicated previously. The estimate is that this now roughly two to three weeks away.

I would also like to elaborate on what it means for the Numba team to validate this patch and why there is a delay in doing so. It will necessitate a full rebuild of the  Numba/llvmlite/LLVM stack and a testing run on all supported versions of Python/Numpy/Operating System/Architecture and a very large selection of combination of these on our (Anaconda-)internal build farm. This work, even though it is largely automated, will still take about two days of exclusive time (for compute) on the farm, even if everything goes well. As I mentioned already, I am unable to re-use the pre-compiled packages for this task as I will need to build the stack from source.

In any case, currently the farm must remain available to continue testing any potential Python 3.9 related bugs (support was merged just a few days ago) and for testing mainline (`master`)  as PRs related to the 0.53 release are merged. To put additional pressure on the farm, there are still a significant number of ongoing upgrades and fixes which need to be put into place to fully prepare the farm to be able to build and cut a release candidate. These are related to upgrading both hardware and software. For example,  the CPU fan on our main build node died during the end-of-year holidays and there were a significant number of delays related to both obtaining a replacement and getting physical access to the hardware as a result of the ongoing global pandemic.

I hope that this summarizes the current situation succinctly for all parties involved.

**Comment by esc** [comment]
> > to port the existing patch from [numba/llvmlite#684](https://github.com/numba/llvmlite/pull/684) to LLVM 11.*
> 
> I'll provide the PR tomorrow, so you have it available when needed.

Excellent, thank you! That will certainly make things easier as and when I start.

**Comment by abebeos** [comment]
> [...]change the relocation model back to static right now. It is not yet clear to me however, if this would have any benefit in terms of usability or performance. Perhaps someone on this thread could share some insight on this?

Would be interested, too. I remember a comment "JIT should use 'static'" within llvm, but without any rationale.

**Comment by edelsohn** [comment]
LLVM 12.0 was just released yesterday, but I fully understand that llvmlite would not jump onto that release immediately.  I'm glad that llvmlite is being updated to LLVM 11.0 and llvmlite can include the backport of the patch.

As far as pic versus static, one would need to test the numba workload on Power in both modes for a definitive answer. Why did numba switch to static?

Is numba continuing to use static on all other architectures? My motivation to fix this for Power is that I want to avoid gratuitous differences for numba on Linux on Power.

**Comment by abebeos** [comment]
> As far as pic versus static, one would need to test the numba workload on Power in both modes for a definitive answer. Why did numba switch to static?

numba: 3990da1b2c5cb743e0fefa50a9a70c7dfdc4b76e - 3990da1b2c5cb743e0fefa50a9a70c7dfdc4b76e (switched to static because llvm did so:

> // JIT codegen should use static relocations by default, since it's
>    // typically executed in process and not relocatable.
https://reviews.llvm.org/D47211#inline-425406

Another switch to "no-pic" here (contains some talk re PowerPC "pic/static")

llvm: https://reviews.llvm.org/D53383 (to get in line with gcc defaults)

> Is numba continuing to use static on all other architectures? My motivation to fix this for Power is that I want to avoid gratuitous differences for numba on Linux on Power.

see https://github.com/numba/numba/blob/1976c664afeb3251bc1246957c4735b1694f84f4/numba/core/codegen.py#L1310-L1318

**Comment by lhames** [comment]
>
> LLVM 12.0 was just released yesterday


LLVM 11.0.1 was just tagged final, but LLVM 12.0 has not been branched yet.
I believe that the plan of record is to branch for LLVM 12.0 on January
26th -- the schedule should be posted on llvm.org soon.


As far as pic versus static, one would need to test the numba workload on
> Power in both modes for a definitive answer. Why did numba switch to
> static?

Would be interested, too. I remember a comment "JIT should use 'static'"
> within llvm, but without any rationale.


I can add a bit of context here:

The advice in early MCJIT days was to use the static relocation model. I
believe the rationale was that static performed better in some cases, and
since the JIT always applies relocations on object load (producing
single-use memory buffers) we may as well take advantage of that.

In Orc (MCJIT's successor) we're aiming to eventually support all
relocation and code models, but starting with the default relocation and
code model of each platform. The rationale is that this path is the best
tested.

My advice would be to keep PIC on Power for now if it's working. If/when
JITLink for Power is developed that would be a good time to reevaluate
these choices.

-- Lang.

**Comment by abebeos** [comment]
> @abebeos Please don't claim the bounty until we have consensus.

@edelsohn, I've addressed everything you've raised back in October, investigated, recreated, reproduced, identified existent fix, tested locally, documented, backported, filed follow-up issues. There is not much more to do for me, and re "consensus", everyone seems to agree that **"keeping it at long-time-working reloc-model=pic"** and **"fix static in upcoming release, reevaluate static-vs-pic"** is the obvious way to go. The question to answer (for the deep PowerPC tool-experts) is "Keep reloc-model=pic, or switch to static for JIT compilation?".

Validating the fix is minimal work. The fix itself has minimal effect (in essence just more nops), exclusively on ppc code. The PowerPC developers, especially the author, could validate it in minutes. It can be essentially validated by anyone, by means of "human logic" and/or "testing with binaries". 

I believe that you should ack the bounty at this point.

(If you want me to do more work in this context (numba/llvmlite/llvm itself), then I'd be happy to discuss follow-up tasks via email)

**Comment by abebeos** [comment]
@edelsohn , my thanks to you and IBM!

@lhames , can you please take a look at https://github.com/abebeos/PowerPC/issues/4? It's about JITLink/PowerPC, I'm mainly interested in trying to get a "fast-forward" assessment for "Stage 2" (you can additionally contact me via abebeos @ lazaridis.com).

**Comment by esc** [comment]
> @edelsohn , my thanks to you and IBM!

May I ask, does this mean, that the bounty was awarded?

**Comment by esc** [comment]
> > @edelsohn , my thanks to you and IBM!
> 
> May I ask, does this mean, that the bounty was awarded?

After some more digging, I will now answer my own question:

<img width="774" alt="Screen Shot 2021-01-21 at 11 25 45" src="https://user-images.githubusercontent.com/41563/105338134-7e4ef000-5bdb-11eb-8eb5-8ae6efc9bdc9.png">

https://www.bountysource.com/teams/numba

So, congrats to all involved that this went "over the line"!

**Comment by abebeos** [comment]
> I will be happy to help with RuntimeDyld reviews though, and would be happy to help Numba port their code to LLVM's newer OrcV2 / JITLink APIs once we have a PPC implementation of JITLink.

@lhames , I'm a bit confused here: can you please clarify as for when then ORCv2-migration must happen?

LLVM Documentation says ORCv1 will be removed in LLVM12:

>>The majority of the ORCv1 layers and utilities were renamed with a ‘Legacy’ prefix in LLVM 8.0, and have deprecation warnings attached in LLVM 9.0. In LLVM 12.0 ORCv1 will be removed entirely.?

source:https://llvm.org/docs/ORCv2.html#transitioning-from-orcv1-to-orcv2

(@esc , possibly open a new issue to track ORCv2 migration)

**Comment by edelsohn** [comment]
@lhames Was ORCv2 already updated for PPC64LE Linux? Is someone else in the community working on it?  It is being removed for LLVM 12.0 with no JIT linking and dynamic symbol resolution for PPC64LE Linux?

+ @nemanjai

**Comment by nemanjai** [comment]
> @lhames Was ORCv2 already updated for PPC64LE Linux? Is someone else in the community working on it? It is being removed for LLVM 12.0 with no JIT linking and dynamic symbol resolution for PPC64LE Linux?
> 
>     * @nemanjai

If there are known issues with PPC64LE that need to be resolved, I'd be happy to help (keep in mind that my familiarity with ORCv2 and JIT's in general is almost non-existent). We have recently implemented PC-Relative addressing in the back end for Power10 compilations, perhaps there is some work involved in making that also work in ORCv2.

**Comment by lhames** [comment]
Hi David, Abebeos,

@lhames , I'm a bit confused here: can you please clarify as for when then
> ORCv2-migration must happen?


I think Numba is using ExecutionEngine / MCJIT right now. If that's the
case it will continue to work as-is in LLVM 12. MCJIT is in maintenance
mode and will be deprecated at some point, but it's also low-maintenance so
we're not in a rush.

If Numba was using OrcV1 then you'll need to switch to OrcV2 before moving
to LLVM 12. You can do this without implementing JITLink for PPC -- OrcV2
can use RuntimeDyld as its underlying JIT linker.

Migrating to OrcV2 from either OrcV1 or MCJIT is usually relatively
painless. Many clients have made the switch already.

 @lhames Was ORCv2 already updated for PPC64LE Linux? Is someone else in
> the community working on it? It is being removed for LLVM 12.0 with no JIT
> linking and dynamic symbol resolution for PPC64LE Linux?


OrcV2 will work with the existing RuntimeDyld support for PPC.

RuntimeDyld will continue to be supported for the time being, but we do
want to deprecate it eventually too: JITLink enables many new features,
provides better error handling, and enables more efficient code models.

If there are known issues with PPC64LE that need to be resolved, I'd be
> happy to help (keep in mind that my familiarity with ORCv2 and JIT's in
> general is almost non-existent). We have recently implemented PC-Relative
> addressing in the back end for Power10 compilations, perhaps there is some
> work involved in making that also work in ORCv2.


Nothing OrcV2 specific, but you'll need to teach RuntimeDyld for PPC (and
JITLink, if one is developed) about any relocations involved in this if
they appear in JIT'd code.

-- Lang.

**Comment by esc** [comment]
Hi @lhames 

> I think Numba is using ExecutionEngine / MCJIT right now. If that's the case it will continue to work as-is in LLVM 12. MCJIT is in maintenance mode and will be deprecated at some point, but it's also low-maintenance so we're not in a rush. If Numba was using OrcV1 then you'll need to switch to OrcV2 before moving to LLVM 12. You can do this without implementing JITLink for PPC -- OrcV2 can use RuntimeDyld as its underlying JIT linker. Migrating to OrcV2 from either OrcV1 or MCJIT is usually relatively painless. Many clients have made the switch already. 

Indeed, Numba is (still) on MCJIT and we are expecting to upgrade to OrcV2 this year. I would appreciate if you could link some examples of successful switches or (even better) link to some documentation regarding the process. Thanks!

V-

**Comment by abebeos** [comment]
@edelsohn , my summary-understanding is: no action is needed for PowerPC.

**Comment by abebeos** [comment]
@edelsohn , my summary-understanding is: no action is needed for PowerPC (action possibly needed when numba starts the migration MCJIT => OrcV2)

**Comment by lhames** [comment]
> Indeed, Numba is (still) on MCJIT and we are expecting to upgrade to OrcV2 this year. I would appreciate if you could link some examples of successful switches or (even better) link to some documentation regarding the process. Thanks!

@esc Julia, Postgres and Clasp are the larger clients that I've worked with on upgrades. There is some documentation on moving from OrcV1 to OrcV2 [1], but it's not particularly helpful in this case as it mostly deals with the differences between those two APIs.

The short version of the MCJIT -> OrcV2 upgrade story is: LLJIT is the new MCJIT. The APIs are different, but the all the concepts from MCJIT have counterparts in LLJIT. You can find several LLJIT usage examples in llvm/examples/OrcV2Examples directory, including a few showing how to use the new C APIs.

I took a quick look at numba and llvmlite -- I think you would need new llvmlite bindings for the APIs in llvm/include/llvm-c/Orc.h, llvm/include/llvm-c/LLJIT.h and llvm/include/llvm-c/OrcEE.h. In numba/numba/core/codegen.py you would just need to use an LLJIT instance to back your JitEngine class, rather than an ExecutionEngine.

That said, my other recommendation after looking at your code would be to wait until the LLVM 13 release at the end of this year: The APIs that you have been living on are super stable at this point, whereas LLJIT still has some churn from version to version. I foresee more during LLVM 13 development as we roll out the new ORC runtime. Without a PPC JITLink implementation to enable the new features an upgrade now might cause pain for limited gain*.

[1] https://github.com/llvm/llvm-project/blob/main/llvm/docs/ORCv2.rst#transitioning-from-orcv1-to-orcv2

* OrcV2 does support concurrent compilation of JIT'd code, lazy compilation, and has better error checking, even when using RuntimeDyld. Whether those alone are useful enough to you to justify the upgrade cost I don't know.

**Comment by esc** [comment]
@lhames that's very helpful, thank you. I'll take this into the developer meeting next week and we shall discuss and re-evaluate our goals and timeline for this year the.

**Comment by stuartarchibald** [comment]
> @lhames that's very helpful, thank you. I'll take this into the developer meeting next week and we shall discuss and re-evaluate our goals and timeline for this year the.

@esc let's move discussion to the tracking ticket: https://github.com/numba/numba/issues/2290 ?

@lhames thanks for the info, much appreciated. I'm going to copy it into the tracking ticket above. Will also comment on there :) Thanks again!

**Comment by esc** [comment]
@stuartarchibald sounds good!
